service/controller/legacy/v29/ebs/ebs.go

```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
 		if err != nil {
 			return microerror.Mask(err)
 		}
+	}
+
+	{
+		i := &ec2.DescribeVolumesInput{
+			VolumeIds: []*string{
+				aws.String(volumeID),
+			},
+		}
+
+		v, err := e.client.DescribeVolumes(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(v.Volumes) == 0 {
+			// Fall through.
+			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("volume %#q not found", volumeID))
+			return microerror.Maskf(executionFailedError, "no volume found")
+		}
+
+		d := v.Volumes[0]
+		if len(d.Attachments) > 1 {
+			// Shouldn't happen; log so we know if it is
+			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found multiple attachment for volume %#q", volumeID))
+		}
+
+		for _, a := range d.Attachments {
+			if a.State != nil {
+				attachmentStatus := *a.State
+				if attachmentStatus != "detached" {
+					return microerror.Maskf(volumeAttachedError, "volume %#q still attached", volumeID)
```

Comment:
We still need the error matching in the calling resource. Otherwise the resource returns an error, gets retried by the `retryresource` and then causes the whole loop to fail. As discussed, we want to match for the error we expect as long as the volume is still attached and cancel the resource gracefully.

Reply:
I'm not familiar with this kind of implementation. I will take a look in the other part of the codebase to see how it is implemented

Reply:
You can see this in https://github.com/giantswarm/aws-operator/blob/master/service/controller/clusterapi/v29/resource/tccpsubnet/create.go where `addSubnetsToContext` returns an error that is matched against in `EnsureCreated` where we then cancel the reconciliation.